File: objc/ZendriveSDKDemo/Model/Trip.m
//
//  Trip.m
//  ZendriveSDKDemo
//
//  Created by Yogesh on 3/7/15.
//  Copyright (c) 2015 Zendrive. All rights reserved.
//

#import "Trip.h"
#import "LocationPoint.h"

#define kStartDateTimestampSecondsKey       @"startDateTimestampSeconds"
#define kEndDateTimestampSecondsKey         @"endDateTimestampSeconds"
#define kAverageSpeedKey                    @"averageSpeed"
#define kDistanceKey                        @"distance"
#define kWaypointsKey                       @"waypoints"
#define kTripStatusKey                      @"tripStatus"

@implementation Trip

- (instancetype)initWithDictionary:(NSDictionary *)dictionary {
    self = [super init];
    if (self) {
        NSNumber *startDateTimestampSeconds = [dictionary objectForKey:kStartDateTimestampSecondsKey];
        if (startDateTimestampSeconds) {
            _startDate = [NSDate dateWithTimeIntervalSince1970:(startDateTimestampSeconds.longLongValue)];
        }

        NSNumber *endDateTimestampSeconds = [dictionary objectForKey:kEndDateTimestampSecondsKey];
        if (endDateTimestampSeconds) {
            _endDate = [NSDate dateWithTimeIntervalSince1970:(endDateTimestampSeconds.longLongValue)];
        }

        NSNumber *averageSpeed = [dictionary objectForKey:kAverageSpeedKey];
        if (averageSpeed) {
            _averageSpeed = averageSpeed.doubleValue;
        }

        NSNumber *distance = [dictionary objectForKey:kDistanceKey];
        if (distance) {
            _distance = distance.doubleValue;
        }

        NSMutableArray *waypoints = [[NSMutableArray alloc] init];
        NSArray *waypointDictioariesArray = [dictionary objectForKey:kWaypointsKey];
        for (NSDictionary *waypointDictionary in waypointDictioariesArray) {
            LocationPoint *locationPoint = [[LocationPoint alloc] initWithDictionary:waypointDictionary];
            [waypoints addObject:locationPoint];
        }
        _waypoints = waypoints;

        _tripStatus = [dictionary objectForKey:kTripStatusKey];
    }
    return self;
}

- (NSDictionary *)toDictionary {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];

    if (self.startDate) {
        [dictionary setObject:@(self.startDate.timeIntervalSince1970) forKey:kStartDateTimestampSecondsKey];
    }
    if (self.endDate) {
        [dictionary setObject:@(self.endDate.timeIntervalSince1970) forKey:kEndDateTimestampSecondsKey];
    }

    [dictionary setObject:@(self.averageSpeed) forKey:kAverageSpeedKey];
    [dictionary setObject:@(self.distance) forKey:kDistanceKey];

    NSMutableArray *waypointDictioariesArray = [[NSMutableArray alloc] init];
    for (LocationPoint *locationPoint in self.waypoints) {
        [waypointDictioariesArray addObject:locationPoint.toDictionary];
    }
    [dictionary setObject:waypointDictioariesArray forKey:kWaypointsKey];

    if (self.tripStatus) {
        [dictionary setObject:self.tripStatus forKey:kTripStatusKey];
    }

    return dictionary;
}
@end

File: swift/ZendriveSDKDemoSwift/Model/Trip.swift
//
//  Trip.swift
//  ZendriveSDKDemoSwift
//
//  Created by amrit on 18/07/19.
//  Copyright © 2019 zendrive. All rights reserved.
//

import Foundation

final class Trip {
    var startDate: Date?
    var endDate: Date?
    var averageSpeed: Double
    var waypoints: [LocationPoint]
    var distance: Double
    var tripstatus: String?
    var vehicleId: String?
    var vehicleTaggingMode: String?

    private let startDateKey = "startDate"
    private let endDateKey = "endDate"
    private let averageSpeedKey = "averageSpeed"
    private let distanceKey = "distance"
    private let waypointsKey = "waypoints"
    private let tripStatusKey = "tripStatus"
    private let vehicleIdKey = "vehicleId"
    private let vehicleTaggingModeKey = "vehicleTaggingMode"



    init() {
        averageSpeed = 0.0
        waypoints  = []
        vehicleId = ""
        vehicleTaggingMode = ""
        distance = 0.0
        tripstatus = ""
    }

    convenience init(dictionary: [AnyHashable : Any]?) {
        self.init()
        if let startDate =  dictionary?[startDateKey] as? Date {
            self.startDate = startDate
        }

        if let endDate = dictionary?[endDateKey] as? Date {
            self.endDate = endDate
        }


        if let averageSpeed = dictionary?[averageSpeedKey] as? Double {
            self.averageSpeed = averageSpeed
        }

        if let distance = dictionary?[distanceKey] as? Double {
            self.distance = distance
        }

        if let waypointDictionaries = dictionary?[waypointsKey] as? [[AnyHashable: Any]] {
            var waypoints:[LocationPoint] = []
            for waypointDict in waypointDictionaries {
                waypoints.append(LocationPoint(dictionary: waypointDict))
            }
            self.waypoints = waypoints
        }

        if let vehicleId = dictionary?[vehicleIdKey] as? String {
            self.vehicleId = vehicleId
        }

        if let vehicleTaggingMode = dictionary?[vehicleTaggingModeKey] as? String {
            self.vehicleTaggingMode = vehicleTaggingMode
        }

        if let tripstatus = dictionary?[tripStatusKey] as? String {
            self.tripstatus = tripstatus
        }

    }

    func toDictionary() -> [AnyHashable : Any] {
        var dictionary: [AnyHashable : Any] = [:]
        if let tripstatus = self.tripstatus {
            dictionary[tripStatusKey] = tripstatus
        }

        if let startDate = self.startDate {
            dictionary[startDateKey] = startDate
        }

        if let endDate = self.endDate {
            dictionary[endDateKey] = endDate
        }

        var waypointDictionaries: [[AnyHashable: Any]] = []
        for waypoint in self.waypoints {
            waypointDictionaries.append(waypoint.toDictionary())
        }

        if let vehicleId = self.vehicleId {
            dictionary[vehicleIdKey] = vehicleId
        }

        if let vehicleTaggingMode = self.vehicleTaggingMode {
            dictionary[vehicleTaggingModeKey] = vehicleTaggingMode
        }

        dictionary[waypointsKey] = waypointDictionaries
        dictionary[distanceKey] = self.distance
        dictionary[averageSpeedKey] = self.averageSpeed
        return dictionary
    }
}
